proposals/csharp-10.0/lambda-improvements.md

```diff
@@ -0,0 +1,121 @@
+```
+
+Attributes on the lambda or lambda parameters will be emitted to metadata on the method that maps to the lambda. (If the lambda does not require a closure class, the lambda is emitted as a method on the containing type. Otherwise the lambda is as a method on the generated closure class.)
+
+## Explicit return type
+An explicit return type may be specified after the parameter list.
+```csharp
+f = () : T => default;              // () : T
+f = x : int => 1;                   // <unknown> : int
+f = (ref int x) : ref int => ref x; // ref int : ref int
+f = static _ : void => { };         // <unknown> : void
+```
+
+Explicit return types are not supported for anonymous methods declared with `delegate { }` syntax.
+```csharp
+f = delegate : int { return 1; };         // syntax error
+f = delegate (int x) : int { return x; }; // syntax error
+```
+
+## Natural type
+A lambda expression has a natural type if the parameters types are explicit and either the return type is explicit or there is a common type from the natural types of all `return` expressions in the body. Otherwise there is no natural type.
+
+A method group has a natural type if the method group contains a single method and the method or reduced extension method has no unbound type parameters.
+
+Lambdas or method groups with natural types can be used as initializers in `var` declarations.
+
+If the lambda or method group has no more than 16 parameters and no return value, and all parameters are passed by value, the natural type will be `delegate void System.Action<P1, ..., Pn>(P1, ..., Pn)` where `P1, ..., Pn` are the lambda parameter types.
+
+If the lambda or method group has no more than 16 parameters and a by-value return type, and all parameters are passed by value, the natural type will be `delegate R System.Func<P1, ..., Pn, R>(P1, ..., Pn)` where `P1, ..., Pn` are the lambda parameter types, and `R` is the lambda return type.
+
+Otherwise the natural type will be a synthesized `internal` anonymous `delegate` type with a signature that matches the lambda or method group.
+
+```csharp
+var f1 = () => default;        // error: no natural type
+var f2 = x => { };             // error: no natural type
+var f3 = x => x;               // error: no natural type
+var f4 = () => 1;              // System.Func<int>
+var f5 = () : string => null;  // System.Func<string>
+```
+
+```csharp
+static void F1() { }
+static void F1<T>(this T t) { }
+static void F2(this string s) { }
+
+var f6 = F1;    // error: multiple methods
+var f7 = "".F1; // System.Action
+var f8 = F2;    // System.Action<string> 
+```
+
+### Anonymous delegate type
+If synthesized delegate types are required, the compiler generates generic anonymous delegate types that are shared across all anonymous delegates in the module that have the same number of parameters and same parameter ref kinds.
+
+The names of the synthesized delegate types and the names of the parameters are unspeakable.
```

Comment:
I'm not sure I understand this. Does this mean that `Delegate.Method.GetParameters()` won't return `ParameterInfo`s with the actual `Name` of the parameters for lambdas requiring synthesized delegate types? The parameter names are important to MapAction(). Or is this referring to some other parameter name?

Reply:
> The parameter names are important to MapAction(). 

Then don't we have a more fundamental problem? If the parameter names matter then any re-use of the `Action` and `Func` types will not work because we can't change the names of them.

Reply:
Considering we already get preserved parameter names when we explicitly cast a method group to the appropriate `Action` or `Func` I assumed this is something that was stored in the `Action` or `Func` instance rather than baked into the types themselves.

Reply:
I think this isn't a problem as the synthesized delegate type may have generic names, but `Delegate.Method.GetParameters()` will still have the correct names.
```e.g.
Func<int,int> foo = (int horse)=>horse;
Console.WriteLine(foo.Method.GetParameters()[0].Name)
```
prints "horse".

Reply:
> I assumed this is something that was stored in the `Action` or `Func` instance rather than baked into the types themselves.

All delegates track the runtime method handle of the target method, which is what you're getting when you reference the `Method` property.  That's how you get the parameter names of the target method instead of the placeholder parameter names of the `Func<...>` delegates themselves, e.g. `arg1`, arg2`, etc.

Reply:
Right. This works fine, we care about the target method, not the delegate parameters.